File: frontend/src/workflow-progress.ts
export type WorkflowId = 'firered' | 'qwen2511_lanpaint' | 'flux2klein_lanpaint'

export type WorkflowProgressState = 'waiting' | 'preparing' | 'running' | 'completed' | 'failed' | 'abandoned'

export interface WorkflowProgressRecord {
  state: WorkflowProgressState
  completed: number
  total: number
  generated: number
  passthrough: number
  first_seconds: number | null
  warm_average_seconds: number | null
  elapsed_seconds: number
  remaining_seconds: number | null
  started_at: string | null
  finished_at: string | null
}

export type WorkflowProgressMap = Partial<Record<WorkflowId, WorkflowProgressRecord>>

export const workflowDisplayNames: Record<WorkflowId, string> = {
  qwen2511_lanpaint: 'Qwen Image 2.1 INT8',
  firered: 'FireRed FP8',
  flux2klein_lanpaint: 'Flux2 Klein + LanPaint',
}

const workflowStateLabels: Record<WorkflowProgressState, string> = {
  waiting: '等待執行',
  preparing: '準備中',
  running: '運行中',
  completed: '已完成',
  failed: '失敗',
  abandoned: '已放棄',
}

export function workflowName(workflow: WorkflowId): string {
  return workflowDisplayNames[workflow]
}

export function workflowStateLabel(state: WorkflowProgressState): string {
  return workflowStateLabels[state]
}

export function formatWorkflowSeconds(seconds: number | null): string {
  if (seconds === null || !Number.isFinite(seconds)) return '未記錄'
  const rounded = Math.max(0, Math.round(seconds))
  const hours = Math.floor(rounded / 3600)
  const minutes = Math.floor((rounded % 3600) / 60)
  const remainder = rounded % 60
  if (hours) return `${hours} 小時 ${minutes} 分 ${remainder} 秒`
  if (minutes) return `${minutes} 分 ${remainder} 秒`
  return `${remainder} 秒`
}

const activeStates = new Set<WorkflowProgressState>(['waiting', 'preparing', 'running'])
const terminalStates = new Set<WorkflowProgressState>(['completed', 'failed', 'abandoned'])

export function firstTimingText(progress: WorkflowProgressRecord): string {
  if (progress.first_seconds !== null) return formatWorkflowSeconds(progress.first_seconds)
  if (activeStates.has(progress.state)) return '待完成'
  if (progress.state === 'completed' && progress.completed === progress.passthrough) return '不需生成'
  return '未記錄'
}

export function warmTimingText(progress: WorkflowProgressRecord): string {
  if (progress.warm_average_seconds !== null) return `${formatWorkflowSeconds(progress.warm_average_seconds)}／張`
  if (activeStates.has(progress.state)) return '待完成'
  if (Math.max(0, progress.completed - progress.passthrough) <= 1) return '—'
  return '未記錄'
}

export function remainingTimingText(progress: WorkflowProgressRecord): string {
  if (terminalStates.has(progress.state)) return '—'
  if (progress.remaining_seconds !== null) return formatWorkflowSeconds(progress.remaining_seconds)
  if (progress.state === 'running') return '估算中'
  if (progress.state === 'waiting' || progress.state === 'preparing') return '待估算'
  return '—'
}

export function friendlyWorkflowText(text: string): string {
  return text
    .replaceAll('qwen2511_lanpaint', workflowDisplayNames.qwen2511_lanpaint)
    .replaceAll('flux2klein_lanpaint', workflowDisplayNames.flux2klein_lanpaint)
    .replaceAll('firered', workflowDisplayNames.firered)
}

File: backend/app/workflow_progress.py
"""Per-workflow progress from runner events; no inferred image-quality checks."""
from __future__ import annotations

import json
import math
from datetime import datetime
from pathlib import Path

from .schemas import WorkflowProgress

WORKFLOW_NAMES = {
    'qwen2511_lanpaint': 'Qwen Image 2.1 INT8',
    'firered': 'FireRed FP8',
    'flux2klein_lanpaint': 'Flux2 Klein + LanPaint',
}


class TimingReader:
    """Consume complete JSON lines only, retaining incomplete writes for next poll."""
    def __init__(self, path: Path):
        self.path = path
        self.offset = 0

    def read(self) -> list[dict]:
        if not self.path.exists():
            return []
        rows = []
        with self.path.open('rb') as handle:
            if self.path.stat().st_size < self.offset:
                self.offset = 0
            handle.seek(self.offset)
            while line := handle.readline():
                if not line.endswith(b'\n'):
                    break
                self.offset = handle.tell()
                try:
                    row = json.loads(line)
                except (ValueError, UnicodeDecodeError):
                    continue
                if isinstance(row, dict):
                    rows.append(row)
        return rows


def update_progress(progress: WorkflowProgress, rows: list[dict], *, completed: int,
                    passthrough: int, black_count: int, elapsed: float, now: str) -> None:
    progress.completed = min(completed, progress.total)
    progress.passthrough = passthrough
    progress.elapsed_seconds = round(max(0, elapsed), 3)
    for row in rows:
        if row.get('event') == 'item_start' and not row.get('empty_mask_passthrough'):
            progress.state = 'running'
            progress.active_started_at = row.get('time') or row.get('started_at')
        if row.get('status') != 'completed' or row.get('empty_mask_passthrough'):
            continue
        seconds = row.get('elapsed_seconds')
        stem = row.get('stem')
        if not isinstance(stem, str) or not isinstance(seconds, (int, float)) or isinstance(seconds, bool):
            continue
        if not math.isfinite(seconds) or seconds < 0:
            continue
        progress.timing_samples.setdefault(stem, float(seconds))
        progress.state = 'running'
        progress.active_started_at = None
    values = list(progress.timing_samples.values())
    progress.generated = len(values)
    progress.first_seconds = values[0] if values else None
    progress.warm_average_seconds = sum(values[1:]) / len(values[1:]) if len(values) > 1 else None
    remaining = max(0, progress.total - black_count - (progress.completed - passthrough))
    progress.remaining_seconds = None
    if remaining == 0:
        progress.remaining_seconds = 0
    elif progress.warm_average_seconds is not None:
        active_elapsed = 0
        if progress.active_started_at:
            try:
                active_elapsed = max(0, (datetime.fromisoformat(now) - datetime.fromisoformat(progress.active_started_at)).total_seconds())
            except (ValueError, TypeError):
                pass
        average = progress.warm_average_seconds
        # Keep remaining queued images in the estimate even if the active image is slow.
        progress.remaining_seconds = round(max(0, average - active_elapsed) + (remaining - 1) * average, 3)
